File: windows/MicroKeys.Core.Tests/MapperTests.cs
using MicroKeys.Core;
using Xunit;

namespace MicroKeys.Core.Tests;

public class MapperTests
{
    private static (Mapper, Recorder) Make(string json)
    {
        var rec = new Recorder();
        return (new Mapper(Config.Parse(json), rec), rec);
    }

    [Fact]
    public void HoldFollowsPressAndRelease()
    {
        var (m, rec) = Make("""{"bindings": {"MIC": {"mode": "hold", "keys": "rctrl+rshift"}}}""");
        m.Handle(new PadEvent.Key("ACT10", 1));
        m.Handle(new PadEvent.Key("ACT10", 1));   // repeat while held: ignored
        Assert.Equal(new[] { "down rctrl+rshift" }, rec.Log);
        Assert.Equal(new[] { "ACT10" }, m.HeldKeys);
        m.Handle(new PadEvent.Key("ACT10", 0));
        m.Handle(new PadEvent.Key("ACT10", 0));
        Assert.Equal(new[] { "down rctrl+rshift", "up rctrl+rshift" }, rec.Log);
        Assert.Empty(m.HeldKeys);
    }

    [Fact]
    public void OtherHalfOfMicCapCountsAsAct10ByDefault()
    {
        var (m, rec) = Make("""{"bindings": {"MIC": {"mode": "hold", "keys": "rctrl+rshift"}}}""");
        var seen = new List<string>();
        m.KeyObserved += (id, act) => seen.Add($"{id} {act}");
        m.Handle(new PadEvent.Key("ACT11", 1));
        Assert.Equal(new[] { "ACT10" }, m.HeldKeys);
        m.Handle(new PadEvent.Key("ACT11", 0));
        Assert.Equal(new[] { "down rctrl+rshift", "up rctrl+rshift" }, rec.Log);
        Assert.Equal(new[] { "ACT10 1", "ACT10 0" }, seen);
    }

    [Fact]
    public void SplitMicKeyRoutesEachHalfSeparately()
    {
        var (m, rec) = Make("""{"options": {"split_mic_key": true}, "bindings": {"ACT10": {"mode": "hold", "keys": "rctrl+rshift"}, "ACT11": "f13"}}""");
        var seen = new List<string>();
        m.KeyObserved += (id, act) => seen.Add($"{id} {act}");
        m.Handle(new PadEvent.Key("ACT11", 1));
        m.Handle(new PadEvent.Key("ACT11", 0));
        Assert.Equal(new[] { "down f13", "up f13" }, rec.Log);
        Assert.Empty(m.HeldKeys);
        m.Handle(new PadEvent.Key("ACT10", 1));
        Assert.Equal(new[] { "ACT10" }, m.HeldKeys);
        m.Handle(new PadEvent.Key("ACT10", 0));
        Assert.Equal(new[] { "down f13", "up f13", "down rctrl+rshift", "up rctrl+rshift" }, rec.Log);
        Assert.Equal(new[] { "ACT11 1", "ACT11 0", "ACT10 1", "ACT10 0" }, seen);
    }

    [Fact]
    public void TapFiresOnDownOnly()
    {
        var (m, rec) = Make("""{"bindings": {"ACT06": "ctrl+shift+4"}}""");
        m.Handle(new PadEvent.Key("ACT06", 1));
        m.Handle(new PadEvent.Key("ACT06", 0));
        Assert.Equal(new[] { "down ctrl+shift+4", "up ctrl+shift+4" }, rec.Log);
    }

    [Fact]
    public void RotationFiresOnAnyAct()
    {
        var (m, rec) = Make("""{"bindings": {"ENC_CW": "up", "ENC_CC": "down"}}""");
        m.Handle(new PadEvent.Key("ENC_CW", 2));
        m.Handle(new PadEvent.Key("ENC_CC", 0));
        Assert.Equal(new[] { "down up", "up up", "down down", "up down" }, rec.Log);
    }

    [Fact]
    public void UnboundKeysDoNothing()
    {
        var (m, rec) = Make("""{"bindings": {"ACT06": "a"}}""");
        m.Handle(new PadEvent.Key("AG00", 1));
        m.Handle(new PadEvent.Joystick(0.1, 0.9));
        Assert.Empty(rec.Log);
    }

    [Fact]
    public void ConfigChangeAndReleaseAllLetGoOfHeldKeys()
    {
        var (m, rec) = Make("""{"bindings": {"MIC": {"mode": "hold", "keys": "ralt"}}}""");
        m.Handle(new PadEvent.Key("ACT10", 1));
        m.Config = new Config();
        Assert.Equal(new[] { "down ralt", "up ralt" }, rec.Log);
        m.Handle(new PadEvent.Key("ACT10", 0));   // stale release: nothing held any more
        Assert.Equal(2, rec.Log.Count);
    }
}

File: macos/Tests/MicroKeysCoreTests/MapperTests.swift
import Foundation
import Testing
@testable import MicroKeysCore

final class Recorder: KeySynthesizing {
    var log: [String] = []
    func press(_ chord: KeyChord) { log.append("down \(chord)") }
    func release(_ chord: KeyChord) { log.append("up \(chord)") }
}

private func mapper(_ json: String) throws -> (Mapper, Recorder) {
    let rec = Recorder()
    return (Mapper(config: try Config.parse(Data(json.utf8)), synthesizer: rec), rec)
}

@Suite struct MapperTests {
    @Test func holdFollowsPressAndRelease() throws {
        let (m, rec) = try mapper(#"{"bindings": {"MIC": {"mode": "hold", "keys": "rctrl+rshift"}}}"#)
        m.handle(.key(id: "ACT10", act: 1))
        m.handle(.key(id: "ACT10", act: 1))   // repeat while held: ignored
        #expect(rec.log == ["down rctrl+rshift"])
        #expect(m.heldKeys == ["ACT10"])
        m.handle(.key(id: "ACT10", act: 0))
        m.handle(.key(id: "ACT10", act: 0))
        #expect(rec.log == ["down rctrl+rshift", "up rctrl+rshift"])
        #expect(m.heldKeys == [])
    }

    @Test func otherHalfOfMicCapCountsAsACT10ByDefault() throws {
        let (m, rec) = try mapper(#"{"bindings": {"MIC": {"mode": "hold", "keys": "rctrl+rshift"}}}"#)
        var seen: [String] = []
        m.onKey = { id, act in seen.append("\(id) \(act)") }
        m.handle(.key(id: "ACT11", act: 1))
        #expect(m.heldKeys == ["ACT10"])
        m.handle(.key(id: "ACT11", act: 0))
        #expect(rec.log == ["down rctrl+rshift", "up rctrl+rshift"])
        #expect(seen == ["ACT10 1", "ACT10 0"])
    }

    @Test func splitMicKeyRoutesEachHalfSeparately() throws {
        let (m, rec) = try mapper(#"{"options": {"split_mic_key": true}, "bindings": {"ACT10": {"mode": "hold", "keys": "rctrl+rshift"}, "ACT11": "f13"}}"#)
        var seen: [String] = []
        m.onKey = { id, act in seen.append("\(id) \(act)") }
        m.handle(.key(id: "ACT11", act: 1))
        m.handle(.key(id: "ACT11", act: 0))
        #expect(rec.log == ["down f13", "up f13"])
        #expect(m.heldKeys == [])
        m.handle(.key(id: "ACT10", act: 1))
        #expect(m.heldKeys == ["ACT10"])
        m.handle(.key(id: "ACT10", act: 0))
        #expect(rec.log == ["down f13", "up f13", "down rctrl+rshift", "up rctrl+rshift"])
        #expect(seen == ["ACT11 1", "ACT11 0", "ACT10 1", "ACT10 0"])
    }

    @Test func tapFiresOnDownOnly() throws {
        let (m, rec) = try mapper(#"{"bindings": {"ACT06": "cmd+shift+4"}}"#)
        m.handle(.key(id: "ACT06", act: 1))
        m.handle(.key(id: "ACT06", act: 0))
        #expect(rec.log == ["down cmd+shift+4", "up cmd+shift+4"])
    }

    @Test func rotationFiresOnAnyAct() throws {
        let (m, rec) = try mapper(#"{"bindings": {"ENC_CW": "up", "ENC_CC": "down"}}"#)
        m.handle(.key(id: "ENC_CW", act: 2))
        m.handle(.key(id: "ENC_CC", act: 0))
        #expect(rec.log == ["down up", "up up", "down down", "up down"])
    }

    @Test func unboundKeysDoNothing() throws {
        let (m, rec) = try mapper(#"{"bindings": {"ACT06": "a"}}"#)
        m.handle(.key(id: "AG00", act: 1))
        m.handle(.joystick(angle: 0.1, distance: 0.9))
        #expect(rec.log == [])
    }

    @Test func configChangeAndReleaseAllLetGoOfHeldKeys() throws {
        let (m, rec) = try mapper(#"{"bindings": {"MIC": {"mode": "hold", "keys": "fn"}}}"#)
        m.handle(.key(id: "ACT10", act: 1))
        m.config = Config()
        #expect(rec.log == ["down fn", "up fn"])
        m.handle(.key(id: "ACT10", act: 0))   // stale release: nothing held any more
        #expect(rec.log.count == 2)
    }
}
